### ISSUE TITLE
Availability revisited

### ISSUE THREAD

**Issue by JirkaChadima** [windingtree/wiki#55]
Before actually implementing the Availability part, we should revise the originally proposed model (available  [here](https://github.com/windingtree/wiki/blob/93f639e9d884883127ab6f99afaf09202f803da3/hotel-data-swagger.yaml#L335) with a little discussion in https://github.com/windingtree/wt-write-api/issues/10) in light of new development and current technical state of the platform.

To recap, the original idea was to:
- have a single dedicated document for availability
- have two sections in the availability document
  - latest snapshot - whole state of the hotel at given time
  - update records - partial changes in availability that happen in between snapshots

The motivation for this division was that the whole snapshot might be costly to generate.

**Things to consider:**
- Should the updates and the snapshot be in the same file?
- Do we actually need the standalone updates?
- What are the update strategies for particular storages?
- Does [wt-updates-api](http://github.com/windingtree/wt-updates-api) affect any of this?
- Do we need more restrictions?
- What will happen if we switch from days/nights to hours or even minutes?

**Comment by JirkaChadima** [comment]
My thoughts and ideas:

> What are the update strategies for particular storages?

I think that if you are a hotel system, in general you have two possibilities:
1. You don't care about the current state in a remote storage, you are flushing there new and *complete* data on every change.
2. You first read the remote data, do something with it (i. e. apply a diff), and then upload it back.

> Do we actually need the standalone updates?

Maybe not. For strategy 1., the assumption is that you can actually compute the current snapshot every time from your own data. For strategy 2., it's not that a big overhead to update the read remote data with some small diff.

I can still see a use case for it when your data is large and you are managing a similar update queue in your own system. In that case it's just a matter of formatting the list of updates and flushing it out. The snapshot gets changed once in a while and the update list is reset.

> Should the updates and the snapshot be in the same file?

It depends. If you use strategy 2, then there's no big difference, you still have to download some data, change it and put it back. The only thing that would differ is the size of the data. If you use strategy 1, then it might make sense if you are only keeping track of changes and not the whole snapshot.

So I think both approaches (single vs. standalone document) are viable. However, if we split this into two documents, we have to keep an eye on data integrity, so the updates don't get mixed up with a wrong snapshot.

> Does wt-updates-api affect any of this?

If we decide to split snapshot and updates into two documents, than yes, because it might produce race conditions. That's why we would have to keep the integrity somehow (hashes, links...).

> Do we need more restrictions?

Absolutely, `noArrival` is not enough. Maybe addition of `noDeparture` would suffice at this time.

> What will happen if we switch from days/nights to hours or even minutes?

Then the `AvailabilityForDay` concept will not work at all. We would probably have to switch to a range-based availability instead of fixed timebox availability. However that would affect the rate plans as well.

The range-based availability might also be an interesting concept that would align more with our proposition of rate plans. Another thing to think about for the future is availability for particular room (with number for example) - that won't really work with this approach if a room type has quantity bigger than one (see the discussion linked to the original post).

---
# Proposition
There are probably many approaches to generating availability data, so we should probably start small and then fiddle with the model to incorporate new features. So I propose the following:

- Let's start without the update records. Let's just assume that everyone can somehow generate the whole snapshot at any time.
- If we later decide to do update records, put them into a separate document and implement some form of integrity verification.
- If necessary, we can deal with the `read-and-update` strategy directly in the wt-write-api.

Thoughts and remarks? @czervenka @hynek-urban @AugustoL

**Comment by hynek-urban** [comment]
@JirkaChadima We've talked about this in person already so I guess it's no surprise that I agree with you, at least on the first point (let's start without the update records).

As far as the other points go, I still think the value of "updates" in the data structure is unclear. It seems to me that its benefits are tiny, if any at all, not justifying the drawbacks at hand (larger amount of data flowing back and forth; more complicated data structure; need of cross-document integrity checks; more complicated data interpretation for all consumers).

**Comment by czervenka** [comment]
@JirkaChadima First of all I agree to start off with snapshots only to
follow KISS principle <https://en.wikipedia.org/wiki/KISS_principle> (nobody
ever wanted anything more).

My thoughts:

1) It is no exception to have 10+ room types per hotel (as room type is a
handy way to present even an arrangement of a room). If the hotel allows
booking two years in advance, the snapshot will have around 7k fields
(10*2*365) which needs to be formatted and written to WT platform with each
new booking (at hotel side) and then parsed and checked for changes on the
OTA side. (A scenario where both sides have a copy of the room type
availability in their databases and the WT platform serves as intermediary
for synchronisation between them). Doing this for each availability change
by snapshots only is thousands time more expensive than just publishing /
receiving update in form: -1 room for 1st - 10th of October. The benefits
seems apparent to me here.
2) I agree that the update log is useless if the PM System has to read
availability from Winding Tree and then calculate a diff to know, what
changed. On the other hand, I can hardly imagine a PMS which have to read
availability from the Winding Tree platform to remember, what the last
state it published was. There is countless of more efficient ways to know
what changed from the last export (e. g. storing a timestamp, a DB index or
even the whole dataset). The worst scenario I can imagine is, that the part
of PMS exporting to WT can read current state but has no access to the
changes (i.e. booking log). In that case, it needs to cache the state, it
exported, and subtract it from the latest state to calculate the diff. But
this operation is still cheaper than formatting the JSON file and saves the
same calculation to each OTA.
3) Data consistency can be achieved easily by snapshot ID (e.g. hash of
snapshot content as @JirkaChadima mentioned) included in the updates. Race
condition should not be a problem here as addition operator is commutative.
The updates can have serial number as well to make sure that the updates
log is applied whole and at the right order.

Anyway thanks for the discussion. Let's keep it simple now and smart later.

**Comment by hynek-urban** [comment]
@czervenka Let me briefly reply to your thought no. 1.

I can clearly see the problem with large snapshots you describe. I don't see how having "updates" in the WT infrastructure solves it, though. Given how the WT platform is designed, having a separate updates document means the publisher would need to 1) download it, 2) append the newest update to the array, 3) write the whole array to the storage again (e.g. to S3). Something similar goes on on the consumer side as well. All this while the "updates" document, in your hypothetical example, has a lot more than 7k fields.

IMO the only solution would be bypassing the WT storage mechanism during the communication alltogether, but that comes with its own set of problems...

**Comment by czervenka** [comment]
You are right, updates make sense only if it is possible to read / write update records one by one. This can be achieved for example by an index file with list of urls to separate update files. This conforms to the style of data format we are already using. 

What I don’t understand yet, why the update records should have more than 7k fields. The number of fields in all updates should be exactly the number of bookings plus the number of cancellations (since last snapshot) which should be far less than number of fields of full snapshot (if the snapshot is done in reasonable intervals). 

RG

**Comment by hynek-urban** [comment]
@czervenka I'm not sure I understand the index file approach... care to elaborate?

As far as the number of fields goes, my reasoning was the following: Let's imagine I have a room type with, say, 40 available rooms on a given day. Twenty of those are booked. The same information that can be expressed as a single number in the snapshot form ("quantity: 20") will be represented as twenty different updates.

I admit this is a bit simplistic and can be mitigated somewhat if the hotel has good discipline and refreshes the snapshot "in reasonable intervals". I'm still not entirely convinced it would result in smaller amount data in "updates" in the general case, though.

**Comment by JirkaChadima** [comment]
> Ad 1)

I think the difference between processing the whole snapshot and processing a single update is important, but not the thing that takes the most time. Communication with the remote storage will take the most time, I/O is always the slowest thing.

I don't really see a difference with using "index file". Yes, you can publish an update in an isolated manner, but then you have to update the index file. It's totally the same thing, only with different data. For reading, it would also be slightly less efficient, because you'd have to get both index and update file. 2 is more than 1.

> Ad 2)

I'm afraid that I don't understand your point. The idea of snapshots is, that they change only once in a while. On the READ side there's no need to download it over and over.

> Ad 3)

The race conditions might actually be a problem, but on a different level. Consider the following scenario - we have update list with `1, 2, 3` records. Then that two bookings happen at the same time. Two threads (A, B) at the same time want to write to the updates list - which is a flat file. But thread B has no idea that there's a thread A. So, for a split second, you get `1, 2, 3, A`, but then you get `1, 2, 3, B` and the A is lost. This is traditionally dealt with with transaction isolation levels, but since we're using flat files, we don't have that luxury.

> The same information that can be expressed as a single number in the snapshot form ("quantity: 20") will be represented as twenty different updates.

I think there are two points of view:
1. current state - then it's obviously cheaper to just throw in quantity: 20 and don't bother with updates
2. changes over time - it might really be expensive to generate the whole snapshot, so writing 20 times more data might actually be cheaper.

Thanks for the discussion, what I take from it so far is, that the snapshot concept is a valid one, so I'll start implementing it. The updates part has a lot of issues, so it's probably best to leave it be for a while. I think we are getting into a premature optimization area.

**Comment by czervenka** [comment]
@hynek-urban <https://github.com/hynek-urban> I thought that we want to
implement real-time availability i.e. publish any change in availability
immediately (no matter if by publishing full snapshot or a partial update
record). I furthermore (silently) expected that the update record can
aggregate 20 changes (as well as snapshot can).

**Comment by JirkaChadima** [comment]
Just for the sake of information not getting lost: We had a face-to-face discussion with @czervenka and @hynek-urban about the implementation details and performance of various solutions of the update part. @czervenka will do a write-up of that, there are more things to consider than one might see.

I am closing this for now and we will re-visit the availability once again when we have some validation of the already existing part from the external users and we know what exactly are we aiming for performance-wise.
